From f9a4c1bb22fe0c296b84a70274ff1e92e043ca98 Mon Sep 17 00:00:00 2001
From: Ranjani Vaidyanathan-RA5478 <Ranjani.Vaidyanathan@freescale.com>
Date: Fri, 22 May 2009 10:33:39 -0500
Subject: [PATCH] ENGR00108842-1 : Enable booting with CPU frequency set in redboot.

Set the CPU working points based on the frequency the CPU is booted which
can be changed using the clock command in redboot.
Also fix the CPUFREQ driver so that the BogoMips are correct when CPU freq
is set to a different value in redboot.

Signed-off-by: Ranjani Vaidyanathan-RA5478 <Ranjani.Vaidyanathan@freescale.com>
---
 arch/arm/plat-mxc/cpufreq.c |   10 +++++++---
 1 files changed, 7 insertions(+), 3 deletions(-)

diff --git a/arch/arm/plat-mxc/cpufreq.c b/arch/arm/plat-mxc/cpufreq.c
index 656622a..9ebc5b0 100644
--- a/arch/arm/plat-mxc/cpufreq.c
+++ b/arch/arm/plat-mxc/cpufreq.c
@@ -123,6 +123,9 @@ static int calc_frequency_khz(int target, unsigned int relation)
 {
 	int i;
 
+	if ((target * 1000) == clk_get_rate(cpu_clk))
+		return target;
+
 	if (relation == CPUFREQ_RELATION_H) {
 		for (i = ARRAY_SIZE(imx_freq_table) - 1; i > 0; i--) {
 			if (imx_freq_table[i].frequency <= target)
@@ -242,10 +245,10 @@ static int __init mxc_cpufreq_driver_init(struct cpufreq_policy *policy)
 	imx_freq_table[i].index = 0;
 	imx_freq_table[i].frequency = CPUFREQ_TABLE_END;
 
-	policy->cur = policy->min = policy->max = clk_get_rate(cpu_clk) / 1000;
+	policy->cur = clk_get_rate(cpu_clk) / 1000;
 	policy->governor = CPUFREQ_DEFAULT_GOVERNOR;
-	policy->cpuinfo.min_freq = cpu_freq_khz_min;
-	policy->cpuinfo.max_freq = cpu_freq_khz_max;
+	policy->min = policy->cpuinfo.min_freq = cpu_freq_khz_min;
+	policy->max = policy->cpuinfo.max_freq = cpu_freq_khz_max;
 
 	arm_lpm_clk = cpu_freq_khz_min * 1000;
 	arm_normal_clk = cpu_freq_khz_max * 1000;
@@ -254,6 +257,7 @@ static int __init mxc_cpufreq_driver_init(struct cpufreq_policy *policy)
 	policy->cpuinfo.transition_latency = 10;
 
 	ret = cpufreq_frequency_table_cpuinfo(policy, imx_freq_table);
+
 	if (ret < 0) {
 		clk_put(cpu_clk);
 		regulator_put(gp_regulator);
-- 
1.5.4.4

